File: frontend/src/lib/api/regional-status.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { api, ApiError } from "./client";
import type { ProjectStatusCategory, ReportStatus } from "./project-status";

// Account Reporting / Geo Reporting — manually authored, period-scoped
// narrative reports (see backend/app/api/v1/endpoints/regional_status.py).
// Same shape as project-status.ts's Project Status hooks, generated once
// here and parameterized by scope ("account" | "geo") instead of duplicated,
// since the two are structurally identical — only the owning entity differs.

export type RegionalScope = "account" | "geo";

export type RegionalStatusReport = {
  id: string;
  account_id?: string;
  geo_id?: string;
  period_id: string;
  status: ReportStatus;
  // Decimal — Pydantic serializes it as a JSON string, not a number, to
  // avoid float precision loss (same convention as projects.ts's project_revenue).
  revenue: string | null;
  onsite_fte: string | null;
  offshore_fte: string | null;
  projects_count: number | null;
  key_accomplishments: string | null;
  upcoming_key_releases: string | null;
  leadership_support_required: string | null;
  created_by: string | null;
  reviewed_by: string | null;
  reviewed_at: string | null;
  review_comment: string | null;
  created_at: string;
  updated_at: string;
};

export type RegionalStatusReportPayload = {
  period_id: string;
  status?: ReportStatus;
  revenue?: string;
  onsite_fte?: string;
  offshore_fte?: string;
  projects_count?: number;
  key_accomplishments?: string;
  upcoming_key_releases?: string;
  leadership_support_required?: string;
};

export type RegionalStatusReportUpdatePayload = Partial<Omit<RegionalStatusReportPayload, "period_id">>;

function basePath(scope: RegionalScope, scopeId: string): string {
  return `/${scope}s/${scopeId}/status-reports`;
}

// One report per reporting period — "latest" 404s until the first one is
// created, which is normal, not an error, so it's swallowed to undefined.
export function useLatestRegionalStatusReport(scope: RegionalScope, scopeId: string | null) {
  return useQuery({
    queryKey: ["regional-status-report-latest", scope, scopeId],
    queryFn: async () => {
      try {
        return await api.get<RegionalStatusReport>(`${basePath(scope, scopeId!)}/latest`);
      } catch (err) {
        if (err instanceof ApiError && err.status === 404) return null;
        throw err;
      }
    },
    enabled: !!scopeId,
  });
}

// Full history — used both to look up whether a report already exists for a
// given period (so the form can PUT instead of re-POSTing into the
// scope_id+period_id unique constraint) and to render the hub's history table.
export function useRegionalStatusReports(scope: RegionalScope, scopeId: string | null) {
  return useQuery({
    queryKey: ["regional-status-reports", scope, scopeId],
    queryFn: () => api.get<RegionalStatusReport[]>(basePath(scope, scopeId!)),
    enabled: !!scopeId,
  });
}

function invalidateRegionalStatusReports(
  queryClient: ReturnType<typeof useQueryClient>,
  scope: RegionalScope,
  scopeId: string | null
) {
  queryClient.invalidateQueries({ queryKey: ["regional-status-report-latest", scope, scopeId] });
  queryClient.invalidateQueries({ queryKey: ["regional-status-reports", scope, scopeId] });
}

export function useCreateRegionalStatusReport(scope: RegionalScope, scopeId: string | null) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (payload: RegionalStatusReportPayload) =>
      api.post<RegionalStatusReport>(basePath(scope, scopeId!), payload),
    onSuccess: () => invalidateRegionalStatusReports(queryClient, scope, scopeId),
  });
}

export function useUpdateRegionalStatusReport(scope: RegionalScope, scopeId: string | null) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ id, payload }: { id: string; payload: RegionalStatusReportUpdatePayload }) =>
      api.put<RegionalStatusReport>(`${basePath(scope, scopeId!)}/${id}`, payload),
    onSuccess: () => invalidateRegionalStatusReports(queryClient, scope, scopeId),
  });
}

// Account Review / Geo Review (for Geo Heads / CXO): approve/reject a
// Submitted report.
export type StatusReportReviewPayload = {
  decision: "Approved" | "Rejected";
  comment?: string;
  reviewed_by?: string;
};

export function useReviewRegionalStatusReport(scope: RegionalScope, scopeId: string | null) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ id, payload }: { id: string; payload: StatusReportReviewPayload }) =>
      api.patch<RegionalStatusReport>(`${basePath(scope, scopeId!)}/${id}/review`, payload),
    onSuccess: () => invalidateRegionalStatusReports(queryClient, scope, scopeId),
  });
}

// --- Account/Geo Status grids (redesign of the 3 free-text sections above,
// plus a new 4th "Key Risks / Issues" section, into per-category
// add/edit/delete registers — mirrors lib/api/project-status.ts's item
// hooks, generalized by scope) ---

export type RegionalStatusItem = {
  id: string;
  account_id?: string;
  geo_id?: string;
  period_id: string;
  category: ProjectStatusCategory;
  description: string;
  created_at: string;
  updated_at: string;
};

export type RegionalStatusItemPayload = {
  period_id: string;
  category: ProjectStatusCategory;
  description: string;
};

export type RegionalStatusItemUpdatePayload = { description: string };

function itemsQuery(
  scope: RegionalScope,
  scopeId: string,
  periodId: string,
  category: ProjectStatusCategory
): string {
  return `/${scope}s/${scopeId}/status-items?period_id=${periodId}&category=${encodeURIComponent(category)}`;
}

export function useRegionalStatusItems(
  scope: RegionalScope,
  scopeId: string | null,
  periodId: string | null,
  category: ProjectStatusCategory
) {
  return useQuery({
    queryKey: ["regional-status-items", scope, scopeId, periodId, category],
    queryFn: () => api.get<RegionalStatusItem[]>(itemsQuery(scope, scopeId!, periodId!, category)),
    enabled: !!scopeId && !!periodId,
  });
}

function invalidateRegionalStatusItems(
  queryClient: ReturnType<typeof useQueryClient>,
  scope: RegionalScope,
  scopeId: string | null,
  periodId: string | null,
  category: ProjectStatusCategory
) {
  queryClient.invalidateQueries({ queryKey: ["regional-status-items", scope, scopeId, periodId, category] });
}

export function useCreateRegionalStatusItem(
  scope: RegionalScope,
  scopeId: string | null,
  periodId: string | null,
  category: ProjectStatusCategory
) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (payload: RegionalStatusItemPayload) =>
      api.post<RegionalStatusItem>(`/${scope}s/${scopeId}/status-items`, payload),
    onSuccess: () => invalidateRegionalStatusItems(queryClient, scope, scopeId, periodId, category),
  });
}

export function useUpdateRegionalStatusItem(
  scope: RegionalScope,
  scopeId: string | null,
  periodId: string | null,
  category: ProjectStatusCategory
) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ id, payload }: { id: string; payload: RegionalStatusItemUpdatePayload }) =>
      api.put<RegionalStatusItem>(`/${scope}s/${scopeId}/status-items/${id}`, payload),
    onSuccess: () => invalidateRegionalStatusItems(queryClient, scope, scopeId, periodId, category),
  });
}

export function useDeleteRegionalStatusItem(
  scope: RegionalScope,
  scopeId: string | null,
  periodId: string | null,
  category: ProjectStatusCategory
) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (id: string) => api.delete(`/${scope}s/${scopeId}/status-items/${id}`),
    onSuccess: () => invalidateRegionalStatusItems(queryClient, scope, scopeId, periodId, category),
  });
}

File: backend/app/models/regional_status.py
import uuid
from datetime import datetime
from decimal import Decimal

from sqlalchemy import DateTime, ForeignKey, Numeric
from sqlalchemy.orm import Mapped, mapped_column

from app.core.db import Base
from app.models.mixins import TimestampColumns, UUIDPrimaryKey


# Account Reporting / Geo Reporting — manually authored, period-scoped
# narrative reports (see db/tables/34_account_geo_status_reports.sql).
# Mirrors ProjectStatusReport exactly, keyed by account_id/geo_id instead.
class AccountStatusReport(Base, UUIDPrimaryKey, TimestampColumns):
    __tablename__ = "account_status_reports"

    account_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("accounts.id", ondelete="CASCADE"))
    period_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("reporting_periods.id"))
    status: Mapped[str]  # Draft, Submitted
    # Key Metrics — captured once per report alongside the narrative tabs.
    revenue: Mapped[Decimal | None] = mapped_column(Numeric)
    onsite_fte: Mapped[Decimal | None] = mapped_column(Numeric)
    offshore_fte: Mapped[Decimal | None] = mapped_column(Numeric)
    projects_count: Mapped[int | None]
    key_accomplishments: Mapped[str | None]
    upcoming_key_releases: Mapped[str | None]
    leadership_support_required: Mapped[str | None]
    created_by: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("users.id"))
    # Review/sign-off by the level above (Geo Head) — set once the report
    # transitions Submitted -> Approved/Rejected.
    reviewed_by: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("users.id"))
    reviewed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    review_comment: Mapped[str | None]


class GeoStatusReport(Base, UUIDPrimaryKey, TimestampColumns):
    __tablename__ = "geo_status_reports"

    geo_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("geos.id", ondelete="CASCADE"))
    period_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("reporting_periods.id"))
    status: Mapped[str]  # Draft, Submitted
    # Key Metrics — captured once per report alongside the narrative tabs.
    revenue: Mapped[Decimal | None] = mapped_column(Numeric)
    onsite_fte: Mapped[Decimal | None] = mapped_column(Numeric)
    offshore_fte: Mapped[Decimal | None] = mapped_column(Numeric)
    projects_count: Mapped[int | None]
    key_accomplishments: Mapped[str | None]
    upcoming_key_releases: Mapped[str | None]
    leadership_support_required: Mapped[str | None]
    created_by: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("users.id"))
    # Review/sign-off by the level above (CXO) — set once the report
    # transitions Submitted -> Approved/Rejected.
    reviewed_by: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("users.id"))
    reviewed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    review_comment: Mapped[str | None]


# Account Reporting / Geo Reporting status grids (see
# db/tables/36_account_geo_status_items.sql) — mirrors ProjectStatusItem
# exactly, keyed by account_id/geo_id instead of project_id.
class AccountStatusItem(Base, UUIDPrimaryKey, TimestampColumns):
    __tablename__ = "account_status_items"

    account_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("accounts.id", ondelete="CASCADE"))
    period_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("reporting_periods.id"))
    category: Mapped[str]  # ProjectStatusCategory
    description: Mapped[str]
    # Account -> Geo rollup (see services/geo_rollup.py).
    account_rollup_status: Mapped[str] = mapped_column(default="Pending")  # RollupStatus
    rolled_up_geo_item_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("geo_status_items.id", ondelete="SET NULL")
    )


class GeoStatusItem(Base, UUIDPrimaryKey, TimestampColumns):
    __tablename__ = "geo_status_items"

    geo_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("geos.id", ondelete="CASCADE"))
    period_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("reporting_periods.id"))
    category: Mapped[str]  # ProjectStatusCategory
    description: Mapped[str]
